.github/workflows/main.yaml

```
@@ -1,36 +1,228 @@
-    name: 'Size ${{ matrix.size }}'
+    name: 'generate ${{ matrix.size }}'
+    needs: prevent_duplicates
     strategy:
-      fail-fast: false
       matrix:
-        # size: [ 1024, 1536, 2048, 3072, 4096, 6144, 8192]
-        size: [ 1024, 1536, 2048]
+        size: [ 1024, 1536,  2048, 3072, 4096 ]
+        split: [2]
+        include:
+          - size: 6144
+            split: 4
+          - size: 7680
+            split: 12
+          - size: 8192
+            split: 24
+
 
     runs-on: ubuntu-latest
     steps:
 
-    # Make sure we don't have multiple jobs
-    - uses: chevah/auto-cancel-redundant-job@v1
-
     - name: Candidate
       run: |
-        ssh-keygen -G candidate-${{ matrix.size }} -b ${{ matrix.size }}
+
+        ssh-keygen -M generate -O bits=${{ matrix.size }} full.1 &
+        PID1=$!
+        ssh-keygen -M generate -O bits=${{ matrix.size }} full.2 &
+        PID2=$!
+
+        # After getting the PID of each run wait for both
+        # and record any failure signaled via a non-zero exit code.
+        FAIL=0
+        wait $PID1 || let "FAIL+=1"
+        wait $PID2 || let "FAIL+=1"
+
+        cat full.2 >> full.1
+
+        # Split as screening a single file on GitHub can't take more than
+        # 6 hours.
+        # We split with 2 digits as 7680 requires 12 chunks.
+        # Split as round-robin as it looks like the first batch alwasy
+        # takes longer to validate.
+        split -d -a 2 -n r/${{ matrix.split }} full.1 g-${{ matrix.size }}.
+
+        exit $FAIL
+
+    - uses: actions/upload-artifact@v2
+      with:
+        name: candidate-${{ matrix.size }}
+        path: g-*
+        retention-days: 1
+
+
+  screen:
+    needs: generate
+    name: 'screen ${{ matrix.size }}'
+    strategy:
+      matrix:
+        size: [ 1024, 1536, 2048, 3072, 4096 ]
+        split: [2]
+        start: [0]
+
+        # For big primes it can take 30 hours on a single CPU.
+        include:
+          - size: 6144
+            split: 4
+            start: 0
+          - size: 6144
+            split: 4
+            start: 2
+
+          - size: 7680
+            split: 12
+            start: 0
+          - size: 7680
+            split: 12
+            start: 2
+          - size: 7680
+            split: 12
+            start: 4
+          - size: 7680
+            split: 12
+            start: 6
+          - size: 7680
+            split: 12
+            start: 8
+          - size: 7680
+            split: 12
+            start: 10
+
+          - size: 8192
+            split: 24
+            start: 0
+          - size: 8192
+            split: 24
+            start: 2
+          - size: 8192
+            split: 24
+            start: 4
+          - size: 8192
+            split: 24
+            start: 6
+          - size: 8192
+            split: 24
+            start: 8
+          - size: 8192
+            split: 24
+            start: 10
+          - size: 8192
+            split: 24
+            start: 12
```

is split into 24, then this needs to run on 12 runners
```suggestion
          - size: 8192
            split: 24
            start: 12
          - size: 8192
            split: 24
            start: 14
          - size: 8192
            split: 24
            start: 16
          - size: 8192
            split: 24
            start: 18
          - size: 8192
            split: 24
            start: 20
          - size: 8192
            split: 24
            start: 22
```